File: cpp/design-patterns/adapter/adapter.cpp
#include <iostream>
#include <memory>

/**
 * @note Adheres Interface Segregation Principle (ISP) in SOLID
 */

/**
 * @brief The Target interface declares a method for the client to interact
 * with.
 */
class Target {
public:
    virtual void request() const = 0;

    /**
     * @brief Virtual destructor to allow proper destruction of derived classes
     */
    virtual ~Target() = default;
};

/**
 * @brief The Adaptee class contains some useful behavior, but its interface is
 * incompatible with the client code.
 */
class Adaptee {
public:
    /**
     * @brief Specific request that Adaptee can perform.
     */
    void specificRequest() const {
        std::cout << "Adaptee's specific request\n";
    }
};

/**
 * @brief The Adapter class implements the Target interface and bridges the gap
 * between Target and Adaptee.
 */
class Adapter : public Target {
private:
    std::unique_ptr<Adaptee> adaptee; /**< Pointer to the Adaptee object */

public:
    /**
     * @brief Construct a new Adapter object with a provided Adaptee.
     * 
     * @param a Pointer to the Adaptee object
     */
    Adapter(std::unique_ptr<Adaptee> a) : adaptee(std::move(a)) {}

    /**
     * @brief Requests a specific action from the Adaptee through the Adapter.
     */
    void request() const override {
        adaptee->specificRequest();
    }
};

/**
 * @brief The main function demonstrates the use of the Adapter pattern.
 * 
 * @return int Program exit status
 */
int main() {
    std::unique_ptr<Adaptee> adaptee = std::make_unique<Adaptee>();
    std::unique_ptr<Adapter> adapter = std::make_unique<Adapter>(std::move(adaptee));
    adapter->request();
    return 0;
}

File: python/design-patterns/adapter/adapter.py
from abc import ABC, abstractmethod

class Target(ABC):
    """Interface for the client to interact with."""
    @abstractmethod
    def request(self):
        """Abstract method to be implemented by concrete classes."""
        pass

class Adaptee:
    """Contains some useful behavior with an incompatible interface."""
    def specific_request(self):
        """Method representing the specific request of the Adaptee."""
        print("Adaptee's specific request")

class Adapter(Target):
    """Adapts the interface of Adaptee to the Target interface."""
    def __init__(self, adaptee: Adaptee):
        """Initialize the Adapter with an instance of Adaptee."""
        self.adaptee = adaptee

    def request(self):
        """Implement the request method to bridge between Target and Adaptee."""
        self.adaptee.specific_request()

def main():
    """Entry point of the program."""
    adaptee = Adaptee()
    adapter = Adapter(adaptee)
    adapter.request()

if __name__ == "__main__":
    main()
